lib/common/src/Amp.php

```diff
@@ -0,0 +1,261 @@
+
+    /**
+     * Boilerplate CSS stylesheet for the <noscript> tag.
+     *
+     * @var string
+     */
+    const BOILERPLATE_NOSCRIPT_CSS = 'body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}';
+
+    /**
+     * Boilerplate CSS stylesheet for Amp4Ads & Amp4Email.
+     *
+     * @var string
+     */
+    const AMP4ADS_AND_AMP4EMAIL_BOILERPLATE_CSS = 'body{visibility:hidden}';
+
+    /**
+     * Amp runtime tag name.
+     *
+     * @var string
+     */
+    const RUNTIME = 'amp-runtime';
+
+    // Amp classes reserved for internal use.
+    const LAYOUT_ATTRIBUTE          = 'i-amphtml-layout';
+    const NO_BOILERPLATE_ATTRIBUTE  = 'i-amphtml-no-boilerplate';
+    const LAYOUT_CLASS_PREFIX       = 'i-amphtml-layout-';
+    const LAYOUT_SIZE_DEFINED_CLASS = 'i-amphtml-layout-size-defined';
+    const SIZER_ELEMENT             = 'i-amphtml-sizer';
+    const INTRINSIC_SIZER_ELEMENT   = 'i-amphtml-intrinsic-sizer';
+
+    /**
+     * Check if a given node is the Amp runtime script.
+     *
+     * The Amp runtime script node is of the form '<script async src="https://cdn.ampproject.org...v0.js"></script>'.
+     *
+     * @param DOMNode $node Node to check.
+     * @return bool Whether the given node is the Amp runtime script.
+     */
+    public static function isRuntimeScript(DOMNode $node)
+    {
+        if (
+            ! self::isAsyncScript($node)
+            || self::isExtension($node)
+        ) {
+            return false;
+        }
+
+        $src = $node->getAttribute(Attribute::SRC);
+
+        if (strpos($src, self::CACHE_ROOT_URL) !== 0) {
+            return false;
+        }
+
+        if (
+            substr($src, -6) !== '/v0.js'
+            && substr($src, -14) !== '/amp4ads-v0.js'
+        ) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if a given node is the Amp viewer script.
+     *
+     * The Amp viewer script node is of the form '<script async
+     * src="https://cdn.ampproject.org/v0/amp-viewer-integration-...js>"</script>'.
+     *
+     * @param DOMNode $node Node to check.
+     * @return bool Whether the given node is the Amp runtime script.
+     */
+    public static function isViewerScript(DOMNode $node)
+    {
+        if (
+            ! self::isAsyncScript($node)
+            || self::isExtension($node)
+        ) {
+            return false;
+        }
+
+        $src = $node->getAttribute(Attribute::SRC);
+
+        if (strpos($src, self::CACHE_HOST . '/v0/amp-viewer-integration-') !== 0) {
```

Comment:
How about using `substr()` as is being used below? No need to search the entire string if it is supposed to start at.

Reply:
See above.